### PR TITLE
[GPU] Fix ShapeOf primitive

### PR DESCRIPTION
**Pull request by okhovan**
Ticket: CVS-124310

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/shape_of.cpp [hunk 1]
```diff
@@ -12,36 +12,33 @@
...
 
     return layout{dt, format::bfyx, out_size};
 }
 
 template<typename ShapeType>
 std::vector<layout> shape_of_inst::calc_output_layouts(shape_of_node const& /*node*/, const kernel_impl_params& impl_param) {
-    auto prim = impl_param.typed_desc<shape_of>();
-
-    data_types output_dt = prim->output_data_types[0].value_or(data_types::i32);
-    if (impl_param.has_fused_primitives()) {
```
Comment (ytorzuk-altran) [inline src/plugins/intel_gpu/src/graph/shape_of.cpp:37]: What is the reason of dropping data type deduction for fused primitives?

Comment (okhovan) [reply src/plugins/intel_gpu/src/graph/shape_of.cpp:37]: It is not dropped, just moved to get_output_data_type() function - see [line 17](https://github.com/openvinotoolkit/openvino/pull/20983/files#diff-6720aa440603913c1ca4e73873fee209d6d337e3cf535c0ea14d981aa8ac9314R17) in this file.

**Comment by ytorzuk-altran** [comment]
PR changes LGTM

File: src/plugins/intel_gpu/src/graph/shape_of.cpp [hunk 2]
```diff
@@ -12,36 +12,33 @@
...
-layout shape_of_inst::calc_output_layout(shape_of_node const& node, kernel_impl_params const& impl_param) {
-    auto prim = impl_param.typed_desc<shape_of>();
-
-    data_types dt = data_types::i32;
-
-    if (prim->output_data_types[0])
-        dt = *prim->output_data_types[0];
-
+namespace {
+data_types get_output_data_type(const kernel_impl_params& impl_param) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/shape_of.cpp:16]: please change this function to static if it is only needed for a local usage.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp [hunk 3]
```diff
@@ -14,18 +14,6 @@ struct shape_of : public primitive_base<shape_of> {
...
-    /// @param output_data_type type of output values. can be i32 and i64.
-    shape_of(const primitive_id& id,
-             const input_info& input,
-             size_t input_rank,
-             const data_types output_data_type,
-             const padding& output_padding = padding())
-        : primitive_base(id, {input}, {output_padding}, {optional_data_type{output_data_type}})
-        , input_rank(input_rank) {}
-
     /// @brief Constructs shape_of primitive.
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp:17]: As far as I know, the original input rank was added for some reason  in this PR https://github.com/openvinotoolkit/openvino/pull/17912, for static shape. @steve-y  Could you recall the prior issue and whether this removal is okay?

Comment (steve-y) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp:17]: @yeonbok This PR has been tested on CVS-110900 for model and functional tests and they worked. And the removal was okay.

File: src/plugins/intel_gpu/src/graph/shape_of.cpp (hunk 2 above)
Comment (okhovan) [reply src/plugins/intel_gpu/src/graph/shape_of.cpp:16]: This function is nested into anonymous namespace, so it is already static.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp (hunk 3 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp:17]: Thanks for confirmation.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**